### ISSUE TITLE
unifi TV: DrmSessionException on TiviMate / other players can't play these playlist

### ISSUE THREAD

**Issue by weareblahs** [weareblahs/unifi-tv#8]
This discussion is for Widevine errors on the playtv@unifi platform. The MPD links are not broken.

**Comment by BudakDesa** [comment]
Then what could be the problem if the MPD link isn't broken. I'm the one that emailing you regarding this issue.

**Comment by weareblahs** [comment]
I still didn't know what's the problem. Maybe it's a problem at TM's side.

**Comment by weareblahs** [comment]
Update: Tested with my own device and it also returns DRM_SESSION_EXCEPTION, and the playtv@unifi App worked well. I'm still investigating the error.

**Comment by BudakDesa** [comment]
Hopefully you'll find what is the main cause and fix it too. How do you find out the mpd link by the way. Is there and kind of tool used to get the link. Would like to learn from you and help with all I can. TQ

**Comment by weareblahs** [comment]
Link found from online sources, started from [this playlist](https://www.yamcode.com/raw/EtX1NPp9rb) (invalid Widevine URL used).

Also, the links were found by using this format:  
``http://1.9.62.158/PLTV/88888888/224/3221227*/3221227*.mpd``  
(asterisks means number ranges, such as 601-700 - it will generate URLs like this:  
``http://1.9.62.158/PLTV/88888888/224/3221227601/3221227601.mpd``
``http://1.9.62.158/PLTV/88888888/224/3221227602/3221227602.mpd``
``http://1.9.62.158/PLTV/88888888/224/3221227603/3221227603.mpd``  
Then I pass these links to IDM by using Batch Download - if the link returns as available (not 404) then I'll skip these link for testing.

Then, I use [this Chrome extension](https://chrome.google.com/webstore/detail/native-mpeg-dash-+-hls-pl/cjfbmleiaobegagekpmlhmaadepdeedn?hl=en) to test it. I tested these links and see what channel is that, then I put the link into the playlist.

(All of the channels are found except the upcoming Tokyo Olympics channels)

**Comment by BudakDesa** [comment]
Is there a possibility that  mpd link been change?

**Comment by weareblahs** [comment]
No, the link isn't changed. Tried other IPs like 1.9.15.104, 1.9.15.100 and 1.9.15.106 and it all returns the same content. There might be a problem with TM's Widevine Server (possibly user-agent check) where it returned this error: ``Widevine Cloud License Server Service certificate request failed. | HTTP Status : 302``  
(302: Redirection)

I will be testing with other IP Addresses starting from 1.9.15.100 to 1.9.15.199 (skipping 1.9.15.100, 1.9.15.104, 1.9.15.106 and 1.9.62.158 - This link is redirected to another 1.9.15.*** server)

**Comment by BudakDesa** [comment]
Will drop you an update. The possibility the link has been changed is there. My friend is working on it now. changed to http://1.9.62.173/

**Comment by weareblahs** [comment]
Test it using ``https://github.com/weareblahs/unifi-tv/raw/main/unifi_prebuilt.m3u8``. If you still recieve DrmSessionException, then comment down below, and I'll change it to PlayReady for testing purposes.

**Comment by linkinstreet** [comment]
Tested just now. Tivimate still showing DrmSessionException with http://1.9.62.173/

**Comment by BudakDesa** [comment]
They might change the channel name too..I'll give an update here.

**Comment by weareblahs** [comment]
Changed the unifi_prebuilt.m3u8 scheme to Microsoft PlayReady (https://ottweb.hypp.tv:81/PlayReady/rightsmanager.asmx) and it shows a black screen on TiviMate (although the device I tested, which is Mi TV Stick supports PlayReady).

**Comment by BudakDesa** [comment]
Yup.. cause it require session...huhuhu

**Comment by elhafizzo** [comment]
playlist is working now.

**Comment by weareblahs** [comment]
Thanks for the update, it's working now. If you met any errors like this, feel free to submit an issue here.

**Comment by aa6my** [comment]
> playlist is working now.

playable ? seems like black screen only for me

device - tx6s android 10

**Comment by khalissuratmen** [comment]
same as me. still black screen.

device - tivo stream 4k

**Comment by linkinstreet** [comment]
@aa6my @khalissuratmen the prebuilt are still using Microsoft PlayReady. Modify that to widevine and your widevine ID and it's good to go

**Comment by weareblahs** [comment]
Modified.

**Comment by weareblahs** [comment]
@aa6my @khalissuratmen Update the playlist using TiviMate. I've updated the prebuilt playlist just now.

**Comment by linkinstreet** [comment]
> @aa6my @khalissuratmen Update the playlist using TiviMate. I've updated the prebuilt playlist just now.

Licence type is still playready from what I can see here

**Comment by weareblahs** [comment]
I just changed the Widevine key, not the ImputStream Adaptive format. Updating now.

**Comment by weareblahs** [comment]
Updated. Check it again.

**Comment by aa6my** [comment]
ok confim working
```
#KODIPROP:inputstream.adaptive.license_type=com.widevine.alpha
#KODIPROP:inputstream.adaptive.license_key=https://www.test.com?device_id=xxxxxxxxxxxxxxxxxxxxxxxxx
#EXTINF:-1 tvg-id="" tvg-logo="" group-title="ETC",TEST
http://x.x.x.x/PLTV/xxxxxxxx/xxx/xxxxxxxxxx/xxxxxxxxxx.mpd
```

**Comment by khalissuratmen** [comment]
@weareblahs thanks a lot! it's work!

**Comment by i-cityzen** [comment]
The issue happen again.

**Comment by weareblahs** [comment]
No issue at my side right now. They might have changed the VUID update time to Daily.

Test it using unifi_prebuilt.m3u8, I've updated the key.

I'll keep monitoring any changes on the playtv@unifi website.

**Comment by weareblahs** [comment]
Update (1:12pm): Seems like my side keeps loading (no DrmSessionException error).

**Comment by weareblahs** [comment]
Update (1:14pm): Found that the website itself now uses a different key every login. Don't log out from playtv@unifi when you're watching on TiviMate.

**Comment by i-cityzen** [comment]
Are u happen to had any issues when login to https://playtv.unifi.com.my/ mine keep popup login time out and appears every time I reload the page.

**Comment by i-cityzen** [comment]
> Are u happen to had any issues when login to https://playtv.unifi.com.my/ mine keep popup login time out or the service is temporarily unavailable and its appears every time I reload the page.

Okay, understand now. That link can't be access using other ISP (mine is Maxis). Need to use TM line only.

**Comment by weareblahs** [comment]
> Are u happen to had any issues when login to https://playtv.unifi.com.my/ mine keep popup login time out or the service is temporarily unavailable and its appears every time I reload the page.

Me too. I'm using unifi Air.

When you're at the "Service Temporarily Unavailable" prompt, you can just simply press Ctrl + Shift + I on your keyboard to access Inspect Element, then go to Application, then... you know the rest of how to find the key.

I tried using my Maxis 4G internet and it can access to playtv@unifi smoothly.

**Comment by i-cityzen** [comment]
> > Are u happen to had any issues when login to https://playtv.unifi.com.my/ mine keep popup login time out or the service is temporarily unavailable and its appears every time I reload the page.
> 
> Me too. I'm using unifi Air.
> 
> When you're at the "Service Temporarily Unavailable" prompt, you can just simply press Ctrl + Shift + I on your keyboard to access Inspect Element, then go to Application, then... you know the rest of how to find the key.
> 
> I tried using my Maxis 4G internet and it can access to playtv@unifi smoothly.

Yup, that's how I do it. Just asking if you also experience the same. Weird tho, I can access it smoothly when connected to Unifi. When using maxis, the popup keep appearing.

**Comment by bahala520** [comment]
Why can't it be played on kodi?

**Comment by weareblahs** [comment]
Kodi has already been added to the "Players That Can't Play This Playlist" section as it can't decrypt Widevine unless it's a plugin*. Even PVR IPTV Simple Client can't handle this playlist.

* Technical stuffs: M3U8 playlist-like files were stored in a plugin with the format ".strm", using the same format as unifi.m3u8.

TL;DR: Due to Widevine, this playlist can't play on Kodi.
